src/Playwright/Core/CDPSession.cs

```diff
@@ -0,0 +1,104 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Playwright.Transport;
+using Microsoft.Playwright.Transport.Channels;
+
+#nullable enable
+
+namespace Microsoft.Playwright.Core;
+
+internal class CDPSession : ChannelOwnerBase, ICDPSession, IChannelOwner<CDPSession>
+{
+    private readonly CDPChannel _channel;
+    private readonly Dictionary<string, Action<JsonElement?>> _eventListeners;
+
+    public CDPSession(IChannelOwner parent, string guid) : base(parent, guid)
+    {
+        _eventListeners = new();
+        _channel = new(guid, parent.Connection, this);
+
+        _channel.CDPEvent += OnCDPEvent;
+    }
+
+    ChannelBase IChannelOwner.Channel => _channel;
+
+    IChannel<CDPSession> IChannelOwner<CDPSession>.Channel => _channel;
+
+    public async Task DetachAsync()
+    {
+        await _channel.DetachAsync();
+    }
+
+    public async Task<JsonElement?> SendAsync(string method, Dictionary<string, object>? args = null)
+    {
+        return await _channel.SendAsync(method, args);
+    }
+
+    private void OnCDPEvent(object sender, CDPChannelEventArgs e)
+    {
+        if (_eventListeners.TryGetValue(e.EventName, out Action<JsonElement?> listener))
+        {
+            listener(e.EventParams);
+        }
+    }
+
+    public void AddEventListener(string eventName, Action<JsonElement?> eventHandler)
```

Comment:
Implementing add/removeEventListener is typically a tedious task with several gotchas and edge conditions (like removing self while dispatching, removing dispatched while dispatching, removing non-dispatched while dispatching, etc.). I wonder if we can stay away from this and offer a different API:

```csharp
session.event("Animation.animationCreated") += (_) => Console.WriteLine("Animation created!");
```

This reads and implements simpler to me.

Reply:
I definitely like that syntax better and explored that for a bit but couldn't find a way to make it work. 

The problem is that the `+=` creates a new object that combines the original handler with the new handler and tries to assign that back to the original. So `A += B` is equivalent to `A = A + B`, but we can't assign to the return value of the method.

The only idea I had was to create a custom indexed class that would allow us to do `session. Event["Animation.animationCreated"]' and automatically generate the event handler with a noop if it didn't exist. We could then `+=` to combine with that noop.

I wasn't too keen to start creating what amounts to a custom dictionary implementation and I was worried that it might be confusing to consumers of the interface to have something that looks kind of like a dictionary but has an odd behavior of being both empty and every `session. Event[""] returns a value. I'm open to it though if everyone else thinks that's the way to go.

Reply:
Found a way to make this work by returning an object that implements a standard C# event allowing consumer to add event listeners in a standard C# way
`client.Event("Network.requestWillBeSent").OnEvent += myEventHandler;`